src/object/obj_class.c

```diff
@@ -270,9 +270,17 @@ daos_oclass_fit_max(daos_oclass_id_t oc_id, int domain_nr, int target_nr,
 	}
 
 	grp_size = daos_oclass_grp_size(&ca);
-	if (ca.ca_grp_nr == DAOS_OBJ_GRP_MAX)
+	if (ca.ca_grp_nr == DAOS_OBJ_GRP_MAX) {
 		ca.ca_grp_nr = max(1, (target_nr / grp_size));
 
+		/* To honor RF setting during failure cases, let's reserve RF
+		 * groups, so if some targets fail, there will be enough replacement
+		 * targets to rebuild, so to avoid putting multiple shards in the same
+		 * domain, which may break the RF setting.
+		 */
+		if (ca.ca_grp_nr > rf_factor)
+			ca.ca_grp_nr -= rf_factor;
```

Comment:
what if the RF==0 but the user is using oclass for an object like EC_*GX?
then group number is not shrunk or adjusted. the container or group RF is just a minimum setting, but objects can have higher level of redundancy themselves.

Reply:
If RF = 0, then it means the user does not "care" the data lose, i.e. it is ok to assign multiple shards in the same domain, IMHO.  

And, if the user really want to use all targets for performance reason,  then he can choose RF = 0 .

Reply:
no that's not the case. one can set rf=0 and individually do redundancy on some object and others not. so we cannot assume: "If RF = 0, then it means the user does not "care" the data lose"

Reply:
anyway this is still better than before at least. let's check for @gnailzenh input on this.

Reply:
"If RF = 0, then it means the user does not "care" the data lose"

Oh, sorry for confusion, "not care data lose" may not be the right word here.  I mean RF = 0, any target exclusion will cause container to be inaccessible, i.e. the user get notification. So it does not care if it will put multiple shards in the same domain.

Reply:
>I also have the same question as Mohamad, if RF=0, I think we can still check ca.ca_resil_degree correct? Or you want to keep this as an interface to try all targets for EC & replicaiton?

Hmm, this patch is for honoring the RF setting, so
1. the user can get notification if there are potential data loss.
2. the data in the container can keep complete as long as the number of failure targets is less than the RF setting.

So if RF is 0, then the user will get notification if any targets excluding happened, and also probably the user want to get maximum performance anyway. IMHO, we do not need reserve the target anyway.

> The last question is, if user sets RC=2 and use EC_16P2_GX, we need to reserve 36 targets, but if user chooses EC_4P2_GX, then this function reserves 12 targets, I'm wondering if we can set an upper bound for the number, so we don't have to reserve too many targets.

Indeed, probably we need set some boundary here.